clients/hadoopfs/src/main/java/io/lakefs/FileSystemTracer.java

```diff
@@ -0,0 +1,353 @@
+     */
+    @Override
+    public void setWorkingDirectory(Path new_dir) {
+        LOG.trace("setWorkingDirectory(new_dir {})", new_dir);
+
+        if (useLakeFSFileSystemRes) {
+            lfsFileSystem.setWorkingDirectory(new_dir);
+        }
+        s3AFileSystem.setWorkingDirectory(translateLakeFSPathToS3APath(new_dir));
+    }
+
+    @Override
+    public Path getWorkingDirectory() {
+        LOG.trace("getWorkingDirectory()");
+
+        Path lakefsRes = lfsFileSystem.getWorkingDirectory();
+        Path s3aRes = s3AFileSystem.getWorkingDirectory();
+        LOG.trace("{}[getWorkingDirectory] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+
+        if (useLakeFSFileSystemRes) {
+            return lakefsRes;
+        }
+        return s3aRes;
+    }
+
+    @Override
+    public boolean mkdirs(Path f, FsPermission permission) throws IOException {
+        LOG.trace("mkdirs(f {}, permission {})", f, permission);
+
+        boolean lakefsRes = lfsFileSystem.mkdirs(f, permission);
+        boolean s3aRes = s3AFileSystem.mkdirs(translateLakeFSPathToS3APath(f), permission);
+        LOG.trace("{}s[mkdirs] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+
+        if (useLakeFSFileSystemRes) {
+            return lakefsRes;
+        }
+        return s3aRes;
+    }
+
+    @Override
+    public FileStatus getFileStatus(Path f) throws IOException {
+        LOG.trace("getFileStatus(f {})", f);
+
+        FileStatus lakefsRes = new FileStatus();
+        FileStatus s3aRes = new FileStatus();
+        Path s3aPath = translateLakeFSPathToS3APath(f);
+        boolean fileNotFoundOnLakeFS = false;
+        boolean fileNotFoundOnLakeS3A = false;
+        try {
+            lakefsRes = lfsFileSystem.getFileStatus(f);
+        } catch (Exception e) {
+            LOG.error("[getFileStatus] Can't get {} file status with lakeFSFileSystem", f);
+            fileNotFoundOnLakeFS = true;
+        }
+        try {
+            s3aRes = s3AFileSystem.getFileStatus(s3aPath);
+        } catch (Exception e) {
+            LOG.error("[getFileStatus] Can't get {} file status with S3AFileSystem", s3aPath);
+            fileNotFoundOnLakeS3A = true;
+        }
+
+        LOG.trace("{}[getFileStatus] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+
+        if (useLakeFSFileSystemRes) {
+            if (fileNotFoundOnLakeFS) {
+                throw new FileNotFoundException("getFileStatus, not found on lakeFS");
+            }
+            return lakefsRes;
+        }
+
+        if (fileNotFoundOnLakeS3A) {
+            throw new FileNotFoundException("getFileStatus, not found on S3A");
+        }
+        Path lfsPath = translateS3APathToLakeFSPath(s3aPath);
+        s3aRes.setPath(lfsPath);
+        return s3aRes;
+    }
+
+    /**
+     * An output stream encapsulating two output streams, one for each file system the tracer uses.
+     */
+    private class TracerOutputTStream extends OutputStream {
```

Comment:
Cool! I was unaware of it :) 
I looked at using it but decided not to because it was easy to implement and using the commons bundle may cause shading problems.